incremental-font-transfer/src/patch_group.rs

```diff
@@ -0,0 +1,1253 @@
+        //   require erroring on this case, and it's resolved by:
+        //   - In the spec algo patches are selected and applied one at a time.
+        //   - Further it specifically disallows re-applying the same URI later.
+        //   - So therefore we de-dup by retaining the particular instance which has the highest selection
+        //     priority.
+
+        let mut full_invalidation: Vec<FullInvalidationPatch> = vec![];
+        let mut partial_invalidation_ift: Vec<PartialInvalidationPatch> = vec![];
+        let mut partial_invalidation_iftx: Vec<PartialInvalidationPatch> = vec![];
+        // TODO(garretrieger): do we need sorted order, use HashMap instead?
+        let mut no_invalidation_ift: BTreeMap<String, NoInvalidationPatch> = Default::default();
+        let mut no_invalidation_iftx: BTreeMap<String, NoInvalidationPatch> = Default::default();
+
+        // Step 1: sort the candidates into separate lists based on invalidation characteristics.
+        for uri in candidates.into_iter() {
+            // TODO(garretrieger): for efficiency can we delay uri template resolution until we have actually selected patches?
+            // TODO(garretrieger): for btree construction don't recompute the resolved uri, cache inside the patch uri object?
+            match uri.encoding() {
+                PatchEncoding::TableKeyed {
+                    fully_invalidating: true,
+                } => full_invalidation.push(FullInvalidationPatch(uri.into())),
+                PatchEncoding::TableKeyed {
+                    fully_invalidating: false,
+                } => {
+                    if Some(uri.expected_compatibility_id()) == ift_compat_id.as_ref() {
+                        partial_invalidation_ift.push(PartialInvalidationPatch(uri.into()))
+                    } else if Some(uri.expected_compatibility_id()) == iftx_compat_id.as_ref() {
+                        partial_invalidation_iftx.push(PartialInvalidationPatch(uri.into()))
+                    }
+                }
+                PatchEncoding::GlyphKeyed => {
+                    if Some(uri.expected_compatibility_id()) == ift_compat_id.as_ref() {
+                        no_invalidation_ift
+                            .insert(uri.uri_string(), NoInvalidationPatch(uri.into()));
+                    } else if Some(uri.expected_compatibility_id()) == iftx_compat_id.as_ref() {
+                        no_invalidation_iftx
+                            .insert(uri.uri_string(), NoInvalidationPatch(uri.into()));
+                    }
+                }
+            }
+        }
+
+        // Step 2 - now make patch selections in priority order: first full invalidation, second partial, lastly none.
+        if let Some(patch) = full_invalidation.into_iter().next() {
+            // TODO(garretrieger): use a heuristic to select the best patch
+            return Ok(CompatibleGroup::Full(patch));
+        }
+
+        let mut ift_selected_uri: Option<String> = None;
```

Comment:
instead of declaring this mut and the assigning to it, would it make sense to just return a tuple from the map?

as in,

```rust
let (red, blue) = iter.next().map(|y| (y.0, y.1)).unzip();
```